packages/gatsby/cache-dir/loader.js

```diff
@@ -141,6 +141,13 @@ export class BaseLoader {
             throw new Error(`not a valid pageData response`)
           }
 
+          if (window.serverData) {
+            // merge inlined serverData for page-data response for initial page
+            // and unset window.serverData
+            jsonPayload.result.serverData = window.serverData
+            window.serverData = null
+          }
+
```

Comment:
This probably also should check wether this is request for initial page. Right now it relies on order (request for page-data of current page being first one, but it's not great guarantee)

Reply:
Another note - this still result in `getServerData` being called twice for initial page. Maybe we can set a header or something on that page-data request so that engines completely skip calling `getServerData` and understand that runtime will merge that in from initial html

Reply:
solved in different way, no longer applies